### ISSUE TITLE
configurable acceptor pools

### ISSUE THREAD

**Issue by ruslandoga** [mtrudel/bandit#199]
Disclaimer: I'm opening this issue after a short exchange of ideas over email with @mtrudel :)

👋

<details>
<summary>This is very much WIP, but I've been playing with an alternative acceptor pool (similar to Elli and <a href="https://github.com/mtrudel/thousand_island/tree/inline_accept">https://github.com/mtrudel/thousand_island/tree/inline_accept</a>) and it shows <i>potentially</i> promising results.</summary>

---

```elixir
Mix.install [{:thousand_island, "~> 1.0.0-pre"}]

defmodule Echo do
  use ThousandIsland.Handler 
  @impl ThousandIsland.Handler
  def handle_data(data, socket, state) do
    ThousandIsland.Socket.send(socket, data)
    {:continue, state}
  end
end

ThousandIsland.start_link(handler_module: Echo, port: 8080)
```

Benchmarking with https://github.com/satori-com/tcpkali

```console
$ tcpkali --message '$' --connections=3 --duration=15s 127.0.0.1:8080
Destination: [127.0.0.1]:8080
Interface lo0 address [127.0.0.1]:0
Using interface lo0 to connect to [127.0.0.1]:8080
Ramped up to 3 connections.
Total data sent:     4480.2 MiB (4697817755 bytes)
Total data received: 4212.2 MiB (4416793336 bytes)
Bandwidth per channel: 1619.841⇅ Mbps (202480.1 kBps)
Aggregate bandwidth: 2354.846↓, 2504.676↑ Mbps
Packet rate estimate: 266156.2↓, 219914.4↑ (3↓, 15↑ TCP MSS/op)
Test duration: 15.005 s.
```

```elixir
Mix.install [{:serv, github: "ruslandoga/serv"}]

defmodule Echo do
  use Serv.Handler

  def handle_data(data, socket, state) do
    with :ok <- :socket.send(socket, data) do
      {:continue, state}
    end
  end
end

Serv.start_link(port: 8000, handler: Echo)
```

```console
$ tcpkali --message '$' --connections=3 --duration=15s 127.0.0.1:8000
Destination: [127.0.0.1]:8000
Interface lo0 address [127.0.0.1]:0
Using interface lo0 to connect to [127.0.0.1]:8000
Ramped up to 3 connections.
Total data sent:     15194.4 MiB (15932446063 bytes)
Total data received: 15180.7 MiB (15918126112 bytes)
Bandwidth per channel: 5659.893⇅ Mbps (707486.6 kBps)
Aggregate bandwidth: 8486.023↓, 8493.657↑ Mbps
Packet rate estimate: 745806.7↓, 740387.9↑ (9↓, 17↑ TCP MSS/op)
Test duration: 15.0064 s.
```

---

</details>

I guess a plan could be:

- [ ] figure out https://github.com/mtrudel/thousand_island/issues/74 and redo that benchmark (with a higher number of connections)
- [ ] more tcp-level benchmarks (e.g. accept rate)
- [ ] (just for the benchmarks) adapt Bandit to use this acceptor pool and see how this performance translates to http / ws
- [ ] identify places in Bandit where the acceptor pool can be configured, and what exactly acceptor pools need to implement for Bandit to work (e.g. suspend  / resume currently seem optional)
- [ ] check with the community if any of that configurability / alternative acceptor pools are actually desired

**Comment by mtrudel** [comment]
tl;dr: this is a great tracking ticket, but should be on Thousand Island and not Bandit. Suggest we open a tracking ticket on Thousand Island with same information.

I don't *think* there's anything to change from Bandit's perspective from this work; the changes should be entirely isolated within Thousand Island, and ought to be transparent to Bandit (other than perhaps some documentation updates to the 'Process Model' sections of the READMEs).

In terms of configuration, Bandit is intentionally opaque to the options that get passed in `thousand_island_options`, so any configuration points we would add to Thousand Island to support / configure this would 'just work' in Bandit, no changes needed.
